### PR TITLE
Add Bing Search plugin

### PR DESCRIPTION
**Pull request by ForestLinSen**
Integrates Bing and Baidu search engines into Auto-GPT based on some users' request
* [add use Bing and Baidu to search](https://github.com/Significant-Gravitas/Auto-GPT-Plugins/issues/35)
* [Searching in other popular search engines](https://github.com/Significant-Gravitas/Auto-GPT/issues/748)
* [有没有高手，改成用百度查找资料，用中文使用，适合中国使用](https://github.com/Significant-Gravitas/Auto-GPT/issues/1508)

### CODE REVIEW

**Comment by hdkiller** [comment]
Hello!

Thank you for your contribution! 

I suggest creating two different plugins, one for baidu and one for bing, instead of bundling them together within one.

**Comment by ForestLinSen** [comment]
> Hello!
> 
> Thank you for your contribution!
> 
> I suggest creating two different plugins, one for baidu and one for bing, instead of bundling them together within one.

Hi! Thank you for taking the time to review my pull request and for your suggestion. My thought for combining them was to make it more convenient for Chinese users, as these are the two main search engines that can be accessed without requiring complicated network configurations there. But I understand your point and I'm open to separating them if it's the best approach for our users.

Hi @ntindle , could you share your thoughts on this matter?

**Comment by ntindle** [comment]
Thanks for the contribution! I'm actually about to head to sleep but a quick note is to separate Bing from Baidu so users only have to enable one plug-in if required

**Comment by ntindle** [comment]
Feel free to do them in the same pr though as they are mostly the same

**Comment by ForestLinSen** [comment]
> Thanks for the contribution! I'm actually about to head to sleep but a quick note is to separate Bing from Baidu so users only have to enable one plug-in if required

Thanks for the suggestion! I've updated the pull request as you and @hdkiller  suggested and only included the Bing search functionality. Also, hope you get some good rest!

File: src/autogpt_plugins/bing_search/__init__.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+    # Return the search results as a JSON string
+    return json.dumps(search_results_list, ensure_ascii=False, indent=4)
+
+
+class Message(TypedDict):
+    role: str
+    content: str
+
+
```
[the remark this answers is not in the document]
Comment (ntindle) [reply src/autogpt_plugins/bing_search/__init__.py:51]: Take a look at email for how that’s handled

**Review by ntindle — CHANGES_REQUESTED** [review]

**Comment by ForestLinSen** [comment]
Hi @ntindle  I have made the changes you suggested to the code. Also during the test I found that it can be difficult for the bot to trigger the `bing_search` even when it's registered. So I updated the `pre_command` so that if the user sets the `SEARCH_ENGINE` variable to "bing" and also sets the `AZURE_API_KEY`, the `pre_command` function will return `bing_search` instead of `google_search`

File: src/autogpt_plugins/bing_search/README.zh.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+3. 设置帐户后，转到 "Keys and Endpoint" 部分。
+4. 从那里复制密钥并将其添加到项目目录中的 .env 文件中。
+5. 将环境变量命名为 `AZURE_API_KEY`。
+
+![Azure Key](./screenshots/azure_api.png)
+
+`.env` 文件示例：
+```
+SEARCH_ENGINE=bing
+AZURE_API_KEY=your_azure_api_key
```
Comment (ntindle) [inline src/autogpt_plugins/bing_search/README.zh.md:29]: Call this BING_API_KEY, each azure service has its own key

**Review by ntindle — CHANGES_REQUESTED** [review]

File: src/autogpt_plugins/bing_search/README.zh.md (hunk 2 above)
Comment (ForestLinSen) [reply src/autogpt_plugins/bing_search/README.zh.md]: Hi! I have updated the env variable name and the README file

**Review by ntindle — APPROVED** [review]

**Pull request merged**